db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/ValueNode.java

```diff
@@ -1512,4 +1522,70 @@ public int getOuterJoinLevel() {
     public void setOuterJoinLevel(int level) {
         return;
     }
+
+    public ValueNode replaceIndexExpression(ResultColumnList childRCL) throws StandardException {
+        // by default, try replace this whole subtree
+        if (childRCL != null) {
+            for (ResultColumn childRC : childRCL) {
+                if (this.equals(childRC.getIndexExpression())) {
```

Comment:
It would be a good idea to use a different method than equals in all of the locations where we compare index expressions, like a wrapper function.  For now, perhaps the wrapper function would call equals, but later, when we want a+1 to be matched equivalent with 1+a, we can simply add the new logic to that method and all the required locations will see the new logic.

Reply:
Yes, this is done in the latest changes in DB-10175 PR (https://github.com/splicemachine/spliceengine/pull/4110). Then I merged branch DB-10175 into this one and adapt the code. A few places need to be tuned because `JavaToSQLValueNode` is special in a sense that its sub-tree structure changes during binding. For this reason, we have to bind index expressions (partially, at least), too. It causes a bit trouble.